test/cpp/end2end/xds/xds_override_host_end2end_test.cc

```diff
@@ -449,34 +482,57 @@ TEST_P(OverrideHostTest, ClusterGoneHostStays) {
-  auto* session_state = stateful_session_per_route.mutable_stateful_session()
-                            ->mutable_session_state();
-  session_state->set_name("envoy.http.stateful_session.cookie");
-  CookieBasedSessionState cookie_config;
-  cookie_config.mutable_cookie()->set_name(kCookieName);
-  session_state->mutable_typed_config()->PackFrom(cookie_config);
-  auto* route = route_config.mutable_virtual_hosts(0)->mutable_routes(0);
-  google::protobuf::Any any;
-  any.PackFrom(stateful_session_per_route);
-  route->mutable_typed_per_filter_config()->emplace(kFilterName, any);
+  *route_config.mutable_virtual_hosts(0)->mutable_routes(0) =
+      BuildStatefulSessionRouteConfig("", kCookieName);
   SetListenerAndRouteConfiguration(balancer_.get(),
                                    BuildListenerWithStatefulSessionFilter(""),
                                    route_config);
-  balancer_->ads_service()->SetEdsResource(BuildEdsResource(EdsResourceArgs(
-      {{"locality0", {CreateEndpoint(0), CreateEndpoint(1)}}})));
+  balancer_->ads_service()->SetEdsResource(
+      BuildEdsResource(EdsResourceArgs({{"locality0", {CreateEndpoint(0)}}})));
   WaitForAllBackends(DEBUG_LOCATION);
   // Get cookie for backend #0.
   auto session_cookie = GetAffinityCookieHeaderForBackend(DEBUG_LOCATION, 0);
-  ASSERT_FALSE(session_cookie.empty());
+  ASSERT_TRUE(session_cookie.has_value());
   // All requests go to the backend we specified
-  CheckRpcSendOk(DEBUG_LOCATION, 5, RpcOptions().set_metadata(session_cookie));
+  CheckRpcSendOk(DEBUG_LOCATION, 5,
+                 RpcOptions().set_metadata({*session_cookie}));
   EXPECT_EQ(backends_[0]->backend_service()->request_count(), 5);
 }
 
+TEST_P(OverrideHostTest, DifferentPerRoute) {
+  constexpr absl::string_view kOverriddenCookie = "overridden-cookie-name";
+  CreateAndStartBackends(1);
+  RouteConfiguration route_config = default_route_config_;
+  *route_config.mutable_virtual_hosts(0)->mutable_routes(0) =
+      BuildStatefulSessionRouteConfig("/grpc.testing.EchoTestService/Echo1",
+                                      "");
+  *route_config.mutable_virtual_hosts(0)->add_routes() =
+      BuildStatefulSessionRouteConfig("/grpc.testing.EchoTestService/Echo2",
+                                      kOverriddenCookie);
+  *route_config.mutable_virtual_hosts(0)->add_routes() =
+      default_route_config_.virtual_hosts(0).routes(0);
+  SetListenerAndRouteConfiguration(
+      balancer_.get(), BuildListenerWithStatefulSessionFilter(), route_config);
+  balancer_->ads_service()->SetEdsResource(
+      BuildEdsResource(EdsResourceArgs({{"locality0", {CreateEndpoint(0)}}})));
+  WaitForAllBackends(DEBUG_LOCATION);
+  // Disabled for "echo1" method
+  auto cookies = GetCookiesForBackend(
+      DEBUG_LOCATION, 0, 1, RpcOptions().set_rpc_method(METHOD_ECHO1));
+  ASSERT_EQ(cookies.empty() ? "" : cookies.front().name, "");
+  // Overridden for "echo2" method
+  cookies = GetCookiesForBackend(DEBUG_LOCATION, 0, 1,
```

Comment:
Please test that RPCs to echo2 honor the override cookie name but not the original cookie name.

Reply:
Done